pkg/migration/migration.go

```diff
@@ -0,0 +1,586 @@
+	"github.com/libopenstorage/operator/pkg/constants"
+	"github.com/libopenstorage/operator/pkg/controller/storagecluster"
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	portworxDaemonSetName  = "portworx"
+	portworxContainerName  = "portworx"
+	migrationRetryInterval = 30 * time.Second
+	phaseAwaitingApproval  = "AwaitingMigrationApproval"
+)
+
+// Handler object that carries out migration of Portworx Daemonset
+// and it's components to operator managed StorageCluster object
+type Handler struct {
+	client client.Client
+	driver storage.Driver
+}
+
+// New creates a new instance of migration handler
+func New(ctrl *storagecluster.Controller) *Handler {
+	return &Handler{
+		client: ctrl.GetKubernetesClient(),
+		driver: ctrl.Driver,
+	}
+}
+
+// Start starts the migration
+func (h *Handler) Start() {
+	ds, err := h.getPortworxDaemonSet()
+	if err != nil {
+		logrus.Errorf("Migration failed to start. %v", err)
+		return
+	} else if ds == nil {
+		logrus.Infof("Migration is not needed")
+		return
+	}
+
+	cluster, err := h.createStorageClusterIfAbsent(ds)
+	if err != nil {
+		logrus.Errorf("Migration failed to create StorageCluster. %v", err)
+		return
+	}
+
+	wait.PollInfinite(migrationRetryInterval, func() (bool, error) {
+		cluster, isNeeded := h.isMigrationNeeded(cluster)
+		if !isNeeded {
+			return true, nil
+		} else if cluster == nil {
+			return false, nil
+		}
+
+		if !h.isMigrationApproved(cluster) {
+			return false, nil
+		}
+
+		if err := h.startMigration(cluster); err != nil {
+			return false, nil
+		}
+		logrus.Infof("Migration completed successfully")
+		return true, nil
+	})
+}
+
+func (h *Handler) createStorageClusterIfAbsent(ds *appsv1.DaemonSet) (*corev1.StorageCluster, error) {
+	clusterName := getPortworxClusterName(ds)
+	stc := &corev1.StorageCluster{}
+	err := h.client.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      clusterName,
+			Namespace: ds.Namespace,
+		},
+		stc,
+	)
+	if err == nil {
+		return stc, nil
+	} else if err != nil && !errors.IsNotFound(err) {
+		return nil, err
+	}
+
+	stc, err = h.createStorageCluster(ds)
+	// TODO: Create dry run specs
+	return stc, err
+}
+
+func (h *Handler) startMigration(cluster *corev1.StorageCluster) error {
+	// TODO: Implement this
+	// 1. Backup existing specs
+	// 2. Migrate daemonset pods
+	// 3. Migrate components
+	// 4. Delete daemonset and components
+	return nil
+}
+
+func (h *Handler) createStorageCluster(
+	ds *appsv1.DaemonSet,
+) (*corev1.StorageCluster, error) {
+	stc := h.constructStorageCluster(ds)
+
+	// TODO: Handle enable stork
+	// TODO: Handle enable autopilot
+	// TODO: Handle enable monitoring
```

Comment:
so besides those components, we could construct everything just from portworx container and existing ds object?

Reply:
There are more items missing which are listed on line 448 - https://github.com/libopenstorage/operator/pull/429/files#diff-f343de058945bc383122601555f6ef2a4988c955c6a702167bf6a99c0775fdccR448